### PR TITLE
Add generic test cases for "q" filters.

### PR DESCRIPTION
**Pull request by whitej6**
# Closes: #5067 
# What's Changed

Implemented missing q filter for ExternalIntegration causing 400 response of form fields sourcing from DynamicModelChoiceField.

# TODO

- [ ] Explanation of Change(s)
- [ ] Added change log fragment(s) (for more information see [the documentation](https://docs.nautobot.com/projects/core/en/stable/development/#creating-changelog-fragments))
- [ ] Attached Screenshots, Payload Example
- [ ] Unit, Integration Tests
- [ ] Documentation Updates (when adding/changing features)
- [ ] Example Plugin Updates (when adding/changing features)
- [ ] Outline Remaining Work, Constraints from Design

### CODE REVIEW

File: nautobot/extras/filters/__init__.py [hunk 1]
```diff
@@ -493,6 +493,12 @@ class Meta:
 
 
 class ExternalIntegrationFilterSet(NautobotFilterSet):
+    q = SearchFilter(
```
Comment (glennmatthews) [inline nautobot/extras/filters/__init__.py:496]: Would be good to add a test case in `nautobot.extras.tests.test_filters` to exercise this.

File: changes/5067.fixed [hunk 2]
```diff
@@ -0,0 +1 @@
+ExternalIntegration missing q search filter causing failure when using DynamicModelChoiceField OR ObjectVar.
```
Comment (glennmatthews) [inline changes/5067.fixed:1]: Perhaps:

```suggestion
Fixed missing `q` (search) filter for ExternalIntegration model.
```

File: nautobot/extras/filters/__init__.py (hunk 1 above)
Comment (whitej6) [reply nautobot/extras/filters/__init__.py:496]: Implemented 2 tests, one to always check if q filter is implemented (I would prefer to find something to skip based on an existing attr instead of hard coding a skip) and a second that validates icontains & iexact by dynamically appending a random string. Not sure I am terribly thrilled with either test but did bubble up some other places to update. Happy to rewrite if you have a suggestion.

File: nautobot/core/testing/filters.py [hunk 3]
```diff
@@ -58,6 +59,14 @@ class FilterTestCase(BaseFilterTestCase):
         queryset = None
         filterset = None
 
+        # dictionary of attributes and lookup expression implemented in q filter
+        # meant to skip attributes that use icontains but cannot append a random string to attribute
```
Comment (glennmatthews) [inline nautobot/core/testing/filters.py:63]: I'm not sure I understand what this comment means - can you perhaps clarify?

Would probably also be good to document that if you're using a `SearchFilter` you probably don't need to define this?

File: nautobot/core/testing/filters.py [hunk 4]
```diff
@@ -155,6 +170,40 @@ def test_tags_filter(self):
...
+            if not self.filterset.declared_filters.get("q"):
+                self.skipTest("`q` filter not implemented")
+
+            obj = self.queryset.first()
+            for attr, lookup_method in self.get_q_filter().items():
+                if lookup_method not in ["icontains", "iexact"]:
+                    # only testing icontains and iexact filter lookups
+                    continue
+                lookup = str(obj.id)
+                if not attr == "id":
```
Comment (glennmatthews) [inline nautobot/core/testing/filters.py:190]: Do we have any models for which `id` **is** part of the q filter? That would seem odd to me.

File: nautobot/core/testing/filters.py [hunk 5]
```diff
@@ -155,6 +170,40 @@ def test_tags_filter(self):
...
+                    if not hasattr(obj, attr):
+                        # only testing direct attrs, does not support nested lookups
+                        continue
```
Comment (glennmatthews) [inline nautobot/core/testing/filters.py:193]: Feels like this could be a potential bug, if the q filter references something that isn't actually an object attribute. Could we instead *specifically* check for nested lookups via something like `"__" in attr`?

File: nautobot/extras/tests/test_filters.py [hunk 6]
```diff
@@ -652,6 +653,9 @@ class GraphQLTestCase(FilterTestCases.NameOnlyFilterTestCase):
     queryset = GraphQLQuery.objects.all()
     filterset = GraphQLQueryFilterSet
 
+    # skip testing query attribute for q filter test
+    q_filter = {"name": "icontains"}
```
Comment (glennmatthews) [inline nautobot/extras/tests/test_filters.py:657]: I don't understand from the comment alone why this is needed here - can you perhaps elaborate?

File: nautobot/extras/tests/test_filters.py [hunk 7]
```diff
@@ -1308,8 +1312,12 @@ def test_peer_id(self):
         params = {"peer_id": [self.devices[2].pk]}
         self.assertEqual(self.filterset(params, self.queryset).qs.count(), 1)
 
+    @skip("Skipping RelationshipAssociation is a through model.")
+    def test_q_filter_exists(self):
+        """Test is not applicable for model."""
+
 
-class RelationshipModelFilterSetTestCase(FilterTestCases.FilterTestCase):
+class RelationshipAssociationFilterSetTestCase(FilterTestCases.FilterTestCase):
```
Comment (glennmatthews) [inline nautobot/extras/tests/test_filters.py:1320]: good spot :)

File: nautobot/extras/tests/test_filters.py [hunk 8]
```diff
@@ -1524,6 +1532,10 @@ def test_regression_distinct_2963(self):
             2,
         )
 
+    @skip("Skipping RelationshipAssociation is a through model.")
```
Comment (glennmatthews) [inline nautobot/extras/tests/test_filters.py:1535]: I think it might be perfectly reasonable to expect through models to also support q filters - thoughts?

File: nautobot/ipam/tests/test_filters.py [hunk 9]
```diff
@@ -59,6 +61,10 @@ class VRFTestCase(FilterTestCases.FilterTestCase, FilterTestCases.TenancyFilterT
     queryset = VRF.objects.all()
     filterset = VRFFilterSet
     tenancy_related_name = "vrfs"
+    q_filter = {
+        "name": "icontains",
+        "description": "icontains",
+    }
```
Comment (glennmatthews) [inline nautobot/ipam/tests/test_filters.py:67]: To make sure I understand - this is being declared so that we *don't* test for the `rd` field that is otherwise part of the searchfilter? Why so?

File: nautobot/ipam/tests/test_filters.py [hunk 10]
```diff
@@ -103,6 +109,9 @@ class RouteTargetTestCase(FilterTestCases.FilterTestCase, FilterTestCases.Tenanc
     queryset = RouteTarget.objects.all()
     filterset = RouteTargetFilterSet
     tenancy_related_name = "route_targets"
+    q_filter = {
+        "description": "icontains",
+    }
```
Comment (glennmatthews) [inline nautobot/ipam/tests/test_filters.py:114]: Why needed? Can we not change the `name` of a RouteTarget?

File: nautobot/users/filters.py [hunk 11]
```diff
@@ -100,6 +100,11 @@ class Meta:
 
 
 class ObjectPermissionFilterSet(BaseFilterSet):
+    q = SearchFilter(
+        filter_predicates={
+            "name": "icontains",
+        }
+    )
```
Comment (glennmatthews) [inline nautobot/users/filters.py:107]: Can we use the `NameSearchFilterSet` mixin here instead?

File: nautobot/extras/filters/__init__.py [hunk 12]
```diff
@@ -590,6 +596,11 @@ class Meta:
 
 
 class ImageAttachmentFilterSet(BaseFilterSet):
+    q = SearchFilter(
+        filter_predicates={
+            "name": "icontains",
+        },
+    )
```
Comment (glennmatthews) [inline nautobot/extras/filters/__init__.py:603]: Can we use the `NameSearchFilterSet` mixin here instead?

**Review by glennmatthews — COMMENTED** [review]
Thanks for the updates!

File: nautobot/core/testing/filters.py [hunk 13]
```diff
@@ -155,6 +170,40 @@ def test_tags_filter(self):
             qs_result = self.queryset.filter(tags=tags[0]).filter(tags=tags[1]).distinct()
             self.assertQuerysetEqualAndNotEmpty(filterset_result, qs_result)
 
+        def test_q_filter_exists(self):
+            """Test the `q` filter exists on a filterset, does not validate the filter works as expected."""
+            self.assertTrue(self.filterset.declared_filters.get("q"))
+
+        def test_q_filter_valid(self):
+            """Test the `q` filter based on attributes in `filter_predicates`."""
+            if not isinstance(self.filterset, SearchFilter):
```
Comment (gsnider2195) [inline nautobot/core/testing/filters.py:179]:
```suggestion
            if not isinstance(self.filterset.declared_filters.get("q"), SearchFilter):
```

File: nautobot/extras/tests/test_filters.py (hunk 7 above)
Comment (gsnider2195) [reply nautobot/extras/tests/test_filters.py:1320]: This is actually an almost identical copy of the previous test case. We should really combine the two.

File: nautobot/core/testing/filters.py (hunk 4 above)
Comment (whitej6) [reply nautobot/core/testing/filters.py:190]: When testing it appears that `id` was dynamically getting inserted even when not declared.

Comment (glennmatthews) [reply nautobot/core/testing/filters.py:190]: Ah right. SearchFilter always includes an `id: iexact` predicate. I'd forgotten that.

File: nautobot/ipam/tests/test_filters.py (hunk 9 above)
Comment (whitej6) [reply nautobot/ipam/tests/test_filters.py:67]: The test's is written to dynamically add a random string to the end of the attr which does not conform to the the data expected.

File: nautobot/extras/tests/test_filters.py (hunk 6 above)
Comment (whitej6) [reply nautobot/extras/tests/test_filters.py:657]: The other field in the q filter is the `query` and the dynamic updates to the fields are not compatible with the graphql query format

File: nautobot/core/testing/filters.py (hunk 5 above)
Comment (whitej6) [reply nautobot/core/testing/filters.py:193]: Do we want to continue on both conditions or replace with `"__" in attr`?

File: nautobot/extras/tests/test_filters.py (hunk 6 above)
Comment (glennmatthews) [reply nautobot/extras/tests/test_filters.py:657]:
```suggestion
    # skip testing "query" attribute for generic q filter test as it's not trivially modifiable
    q_filter = {"name": "icontains"}
```

**Comment by gsnider2195** [comment]
@whitej6 I merged the fix to the ExternalIntegration q filter in #5181 so this PR can be changed to just implementing the unittest and related changes.

File: nautobot/core/testing/filters.py [hunk 14]
```diff
@@ -155,6 +164,76 @@ def test_tags_filter(self):
...
+            field = obj._meta.get_field(filter_field_name)
+
+            if isinstance(field, ManyToOneRel) or isinstance(field, ManyToManyRel):
+                rel_obj = getattr(obj, filter_field_name).first()
+            else:
+                rel_obj = getattr(obj, filter_field_name)
+
+            if not rel_obj:
+                # Raise an error stating no instance found to test on
+                # ?: Should we require an instance to be available or just skip
```
Comment (timizuoebideri1) [inline nautobot/core/testing/filters.py:188]: ?

File: nautobot/extras/tests/test_filters.py (hunk 8 above)
Comment (timizuoebideri1) [reply nautobot/extras/tests/test_filters.py]: What should be the filter predicate for `RelationshipAssociation` q filter?

File: nautobot/dcim/filters/__init__.py [hunk 15]
```diff
@@ -1570,7 +1570,7 @@ def filter_virtual_ip(self, queryset, name, value):
         return queryset.filter(virtual_ip__in=ip_queryset)
 
 
-class InterfaceRedundancyGroupAssociationFilterSet(BaseFilterSet, NameSearchFilterSet):
```
[the remark this answers is not in the document]
Comment (glennmatthews) [reply nautobot/dcim/filters/__init__.py:1573]: Good catch. Any analysis on why this wasn't failing before?

File: nautobot/extras/tests/test_filters.py (hunk 8 above)
Comment (glennmatthews) [reply nautobot/extras/tests/test_filters.py]: Mabye something like `relationship__label, relationship__key`?

File: nautobot/circuits/tests/test_filters.py [hunk 16]
```diff
@@ -15,7 +15,7 @@
 from nautobot.extras.models import Role, Status, Tag
 
 
-class ProviderTestCase(FilterTestCases.NameOnlyFilterTestCase):
+class ProviderTestCase(FilterTestCases.FilterTestCase):
```
Comment (glennmatthews) [inline nautobot/circuits/tests/test_filters.py:18]: Why this change? Provider has a name field, right?

File: nautobot/core/testing/filters.py [hunk 17]
```diff
@@ -155,6 +165,116 @@ def test_tags_filter(self):
             qs_result = self.queryset.filter(tags=tags[0]).filter(tags=tags[1]).distinct()
             self.assertQuerysetEqualAndNotEmpty(filterset_result, qs_result)
 
+        def test_q_filter_exists(self):
+            """Test the `q` filter exists on a filterset, does not validate the filter works as expected."""
+            self.assertTrue(self.filterset.declared_filters.get("q"))
```
Comment (glennmatthews) [inline nautobot/core/testing/filters.py:170]: Possibly?

```suggestion
            self.assertIn("q", self.filterset.declared_filters)
```

File: nautobot/core/testing/filters.py [hunk 18]
```diff
@@ -155,6 +165,116 @@ def test_tags_filter(self):
...
+
+        def _assert_nested_q_filter(self, obj, model_field_name, filter_field_name, lookup_method):
+            """
+            Asserts the validity of a nested filter in the queryset.
+
+            Args:
+                obj: The object on which the filter is applied.
+                model_field_name: The name of the field in the model.
+                filter_field_name: The name of the field being filtered.
+                lookup_method: The method used for the lookup.
```
Comment (glennmatthews) [inline nautobot/core/testing/filters.py:186]: Mind adding a concrete example of this method being called to the docstring? Same request on `_assert_non_nested_q_filter`.

File: nautobot/core/testing/filters.py [hunk 19]
```diff
@@ -155,6 +165,116 @@ def test_tags_filter(self):
...
+
+            if isinstance(field, (ManyToOneRel, ManyToManyRel, ManyToManyField)):
+                rel_obj = getattr(obj, filter_field_name).first()
+            else:
+                rel_obj = getattr(obj, filter_field_name)
+
+            if "__" not in nested_field:
+                self._assert_non_nested_q_filter(rel_obj, model_field_name, nested_field, lookup_method)
+            else:
+                self._assert_nested_q_filter(rel_obj, model_field_name, nested_field, lookup_method)
```
Comment (glennmatthews) [inline nautobot/core/testing/filters.py:199]: Maybe consider doing this with a `while` loop instead of recursion? Not blocking.

File: nautobot/core/testing/filters.py [hunk 20]
```diff
@@ -155,6 +165,116 @@ def test_tags_filter(self):
...
+                raise ValueError(
+                    f"Cannot find valid test data for {queryset.model.__name__} with params {filter_params}"
+                )
+            return queryset
+
+        def test_q_filter_valid(self):
+            """Test the `q` filter based on attributes in `filter_predicates`."""
+
+            if not isinstance(self.filterset.declared_filters.get("q"), SearchFilter):
+                self.skipTest("`q` filter is not a SearchFilter")
```
Comment (glennmatthews) [inline nautobot/core/testing/filters.py:254]: IMO this should be a failure of the test. Though I recognize that may cause apps to start failing upstream testing.

File: nautobot/core/testing/filters.py [hunk 21]
```diff
@@ -155,6 +165,116 @@ def test_tags_filter(self):
...
+
+            if not self.filterset.declared_filters.get("q"):
+                self.skipTest("`q` filter not implemented")
```
Comment (glennmatthews) [inline nautobot/core/testing/filters.py:257]: Maybe this should be a failure too?

File: nautobot/extras/tests/test_filters.py [hunk 22]
```diff
@@ -1679,6 +1591,10 @@ def test_secret_type(self):
         params = {"secret_type": [SecretsGroupSecretTypeChoices.TYPE_PASSWORD]}
         self.assertEqual(self.filterset(params, self.queryset).qs.count(), 2)
 
+    @skip("Skipping SecretsGroupAssociation is a through model.")
+    def test_q_filter_exists(self):
+        """Test is not applicable for model."""
```
Comment (glennmatthews) [inline nautobot/extras/tests/test_filters.py:1596]: Probably could add a q filter that checks `secrets_group__name`, `secret__name`, etc.?

File: changes/5067.added [hunk 23]
```diff
@@ -0,0 +1 @@
+Added `q` (search) filter to `ProviderFilterSet`, `ProviderNetworkFilterSet`, `ImageAttachmentFilterSet` and `ObjectPermissionFilterSet`.
```
Comment (glennmatthews) [inline changes/5067.added:1]: Let's document the new generic test cases as an addition as well.

I wonder, should we add the tests in `next` rather than `develop` just given the potential impact to apps?

File: nautobot/core/testing/filters.py (hunk 20 above)
Comment (timizuoebideri1) [reply nautobot/core/testing/filters.py:254]: I don't believe `SearchFilter` would be very usefull in some `FilterSets`, such as `IPAddressFIlterset` and `PrefixFIlterset`.

File: nautobot/dcim/filters/__init__.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (timizuoebideri1) [reply nautobot/dcim/filters/__init__.py:1573]: Maybe because there was no generic `q` filter test case

File: nautobot/core/testing/filters.py (hunk 20 above)
Comment (glennmatthews) [reply nautobot/core/testing/filters.py:263]: Fair point, I hadn't thought of that. Is there a subset of `Filter` types that we *could* assert that includes those cases?

File: changes/5067.added [hunk 24]
```diff
@@ -0,0 +1,2 @@
+Added `q` (SearchFilter) filter to all filterset missing it.
+Added two generic test cases for `q` filter`test_q_filter_exists` and `test_q_filter_valid`.
```
Comment (glennmatthews) [inline changes/5067.added:2]:
```suggestion
Added two generic test cases for `q` filter: `test_q_filter_exists` and `test_q_filter_valid`. 
```

File: nautobot/core/testing/filters.py (hunk 20 above)
Comment (timizuoebideri1) [reply nautobot/core/testing/filters.py:263]: There are currently test cases for each of these `q` fields in their respective Test classes.
